torch/_functorch/aot_autograd.py

```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
     # TODO: we should kill this
     # (need to default it to not break internal)
     is_train: bool = False
+    # Whether we are tracing through subclasses
+    # (ideally should remove this)
+    is_subclass: bool = False
 
     num_symints_saved_for_bw: Optional[int] = None
 
     def __post_init__(self):
         mutated_inp_indices = [
-            i for i, m in enumerate(self.input_info) if m.mutates_metadata or m.mutates_data
+            i for i, m in enumerate(self.input_info)
+            if m.mutates_data or m.mutates_metadata
         ]
-        # pre-compute the indices of the inputs that are mutated.
-        # When keep_input_mutations is set, we don't need to worry about our epilogue
-        # handling data-only mutations, because we keep them directly in the graph.
+
         mutated_inp_runtime_indices = [
-            i for i, m in enumerate(self.input_info) if m.mutates_metadata or (not self.keep_input_mutations and m.mutates_data)
+            i for i, m in enumerate(self.input_info)
+            if _check_if_mutation_will_be_returned_in_runtime(m, self.keep_input_mutations, self.is_train) and not self.is_subclass
+        ]
+
+        mutated_graph_handled_indices = [
+            i for i, m in enumerate(self.input_info)
+            if _check_if_mutation_can_be_handled_in_fw_graph(m, self.keep_input_mutations, self.is_train) and not self.is_subclass
```

Comment:
We just established that there is a ternary condition, but actually the indices here are not reflecting the ternary; instead mutated_inp_indices is everything, mutated_inp_runtime_indices is MUTATION_OUTSIDE_GRAPH, and mutated_graph_handled_indices is MUTATION_INSIDE_GRAPH